src/Compilers/CSharp/Portable/Syntax/InternalSyntax/BinaryExpressionSyntax.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax
+{
+    internal partial class BinaryExpressionSyntax
```

Comment:
It depends how much we are willing to complicate the system. Recursion on the left for binary expressions is very common. Note, we do not try to specially handle recursion on the right, because deep recursion on the right is not that common. The same goes for recursion for other nodes. It would probably be good to have a way to efficiently traverse a tree without using recursion, but this is definitely out of scope for what I am trying to accomplish in this change.